Administrative Item Edit

LotteHofstede opened the pull request:
This PR depends on #355.
This PR closes #363.

When logged on as an administrator, a user can now edit an item's metadata.

The edit metadata page can be found at `[dspace-url]/items/[item-uuid]/edit/metadata`.

On this page you will be able to add, delete and edit metadata fields. The changes you make will be stored in the ngrx store - even if you don't immediately save them - and will be available as long as you don't refresh the page.

### Add
To add a new metadata field, click the large green `+ Add` button at the top of the page. A new (green) row will be added to the metadata table which you can edit.

### Remove
To remove an existing metadata field, click the red `trash` icon in the last column of the field you would like to remove. The row of the field you removed will turn red.
*You can't edit or remove an already removed field.*

### Edit
To edit an existing metadata field, click the blue `edit` icon in the last column of the field you would like to edit. All fields will turn into an input field which you can edit. Whenever you make an actual change, the row will turn orange.
*This `edit` button will only be when the field is not in edit mode.*

### Close edit mode
To close edit mode and display a field without the input fields, click the green `check` icon.
*This `check` button will only be available when the field is in edit mode.*

### Discard changes of a single field
To undo all changes made to a single field, click the orange `undo` icon in the last column of the field to which changes were made. 
An added field will disappear, a removed or editted field will lose its red/orange background color and reset to its initial value.
*Note that the `undo` icon will only be enabled when there are changes to undo.*

### Discard all changes to all fields
To discard all changes you made, click the orange 'Discard' button at the top of the page.
This will immediately remove all current changes.
When clicking the discard button, you will be shown a notification stating that you discarded all changes, but that you can still undo this by clicking the `Undo` button at the top of the page.*
When clicking the undo button at the top of the page, all your changes will be reinstated.
When you don't click the `Undo`button, your changes will be permanently removed when the notification disappears.

*Note that you can only use the `Discard` button when you made actual changes to the metadata. 
The `Undo` button will only be available when there are changes to undo and the notification is still present.*

### Save
To save all the changes you made to the item, click the save button. This will send an actual request to the backend and update the item.
*The `save` button will only be enabled when there are changes to save.*

### `MetadataField` input - Help & Validation
To make sure you cannot enter invalid data, validation was added to the `MetadataField` input field.
When entering something into this field, a dropdown with metadata fields - that match your input - will be shown from which you can choose.
If the user still enters a non-existing or invalid metadata field, an error message will be shown and you will not be able to save your changes.

### General clean up / refactoring
I got rid of al the occurrences of TNormalized in the DataServices (and the RemoteDataBuildService). Each TDomain has a single TNormalized, so carrying around both seemed unnecessary. In my opinion this makes the DataServices and RemoteDataBuildService easier to understand and maintain.


**) Note that ideally we would have liked this `Undo` button to be part of the notification, but because of https://github.com/DSpace/dspace-angular/issues/357 this is not possible yet.*

LotteHofstede commented on src/app/+item-page/edit-item-page/item-metadata/edit-in-place-field/edit-in-place-field.component.html:
```diff
@@ -0,0 +1,71 @@
+<!--{{metadata?.uuid}}-->
```
Removed comment

LotteHofstede commented on src/app/+item-page/edit-item-page/item-metadata/item-metadata.component.ts:
```diff
@@ -0,0 +1,231 @@
+   */
+  private discardTimeOut: number;
+
+  private notitifactionPrefix = 'item.edit.metadata.notifications.';
```
Fixed typo, added TypeDoc

LotteHofstede commented on src/app/core/registry/registry.service.ts:
```diff
@@ -167,6 +173,41 @@ export class RegistryService {
     return this.rdb.toRemoteDataObservable(requestEntryObs, payloadObs);
   }
 
+  public getAllMetadataFields(pagination?: PaginationComponentOptions): Observable<RemoteData<PaginatedList<MetadataField>>> {
```
Added TypeDoc

LotteHofstede commented on src/app/shared/utils/auto-focus.directive.ts:
```diff
@@ -0,0 +1,21 @@
+@Directive({
+  selector: '[dsAutoFocus]'
+})
+export class AutoFocusDirective implements AfterViewInit {
```
Added TypeDoc

artlowel commented on src/app/core/registry/registry.service.ts:
```diff
@@ -495,4 +556,19 @@ export class RegistryService {
+
+  queryMetadataFields(query: string): Observable<RemoteData<PaginatedList<MetadataField>>> {
+    /**
+     * This should come directly from the server in the future
```
I've talked this over with @benbosman, and we concluded that this type of search is much easier to do in the UI than it would be in the backend. So we decided to keep it here. We'll remove this comment.

tdonohue replied:
Sounds good, thanks!

LotteHofstede commented:
@tdonohue and @paulo-graca Thanks for your feedback. I added the missing TypeDoc, fixed the typo and made sure the notification and Undo-button now show for 10 seconds. I also made this timeout easily configurable in [the environment configuration](https://github.com/DSpace/dspace-angular/blob/27180c5960ac4e4f163899f38c0471b1a5bf8244/config/environment.default.js#L85).
